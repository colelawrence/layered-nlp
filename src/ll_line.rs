mod as_tokens;
mod display;
mod ll_selection;
pub(crate) mod x;

pub(crate) use ll_selection::LLSelection;

use unicode_width::UnicodeWidthStr;

use crate::type_bucket::{self, AnyAttribute};
use crate::type_id_to_many::TypeIdToMany;
pub use display::LLLineDisplay;
use std::fmt::Write;
use std::{collections::HashMap, rc::Rc};
use x::XMatchNext;
pub use x::{Attr, AttrEq};

#[derive(Clone, Debug, PartialEq)]
pub enum TextTag {
    NATN,
    PUNC,
    SYMB,
    SPACE,
    WORD,
}

#[derive(Debug)]
pub enum LToken {
    Text(String, TextTag),
    /// TODO: something more interesting
    Value,
}

#[derive(Debug)]
pub struct LLToken {
    pub(crate) token_idx: usize,
    // token span position (not token index)
    pub(crate) pos_starts_at: usize,
    // token span position (not token index)
    pub(crate) pos_ends_at: usize,
    pub(crate) token: LToken,
}

/// (starts at, ends at) token indexes
type LRange = (usize, usize);

struct LLLineAttrs {
    // "bi-map" / "tri-map"
    ranges: TypeIdToMany<LRange>,
    /// match_forwards uses [LLCursor::end]
    starts_at: Vec<TypeIdToMany<LRange>>,
    /// match_backwards uses [LLCursor::start]
    ends_at: Vec<TypeIdToMany<LRange>>,
    values: HashMap<LRange, type_bucket::TypeBucket>,
}

/// TODO: insert [TextTag]s into the ranges
pub struct LLLine {
    // how much do we actually need of the original Vec if much of the data is put into the bi-map?
    ll_tokens: Vec<LLToken>,
    attrs: LLLineAttrs,
}

impl LLLine {
    pub(crate) fn new(ll_tokens: Vec<LLToken>) -> Self {
        let mut starts_at: Vec<TypeIdToMany<LRange>> = Default::default();
        let mut ends_at: Vec<TypeIdToMany<LRange>> = Default::default();
        for _ in ll_tokens.iter() {
            starts_at.push(Default::default());
            ends_at.push(Default::default());
        }
        let mut attrs = LLLineAttrs {
            ranges: Default::default(),
            starts_at,
            ends_at,
            values: Default::default(),
        };

        for (token_idx, ll_token) in ll_tokens.iter().enumerate() {
            match &ll_token.token {
                LToken::Text(text, tag) => {
                    if text.chars().count() == 1 {
                        // insert char automatically if just one char
                        attrs.insert((token_idx, token_idx), text.chars().next().unwrap());
                    }
                    // insert TextTag automatically
                    attrs.insert((token_idx, token_idx), tag.clone());
                }
                LToken::Value => {
                    // nothing to do...
                }
            }
        }

        LLLine { attrs, ll_tokens }
    }

    pub fn run<R>(mut self, recognizer: &R) -> Self
    where
        R: Resolver,
    {
        let ll_line = Rc::new(self);

        let assignments = recognizer.go(LLSelection {
            ll_line: ll_line.clone(),
            start_idx: 0,
            end_idx: ll_line.ll_tokens().len() - 1,
        });

        self = Rc::try_unwrap(ll_line)
            .map_err(drop)
            .expect("there is no other Rc currently");

        // caches currency symbols
        for LLCursorAssignment {
            start_idx,
            end_idx,
            value,
        } in assignments
        {
            self.attrs.insert((start_idx, end_idx), value);
        }

        self
    }
    pub(crate) fn add_any_attrs(
        &mut self,
        start_idx: usize,
        end_idx: usize,
        attrs: Vec<AnyAttribute>,
    ) {
        let range = (start_idx, end_idx);

        for attr in attrs {
            self.attrs
                .starts_at
                .get_mut(start_idx)
                .expect("has initial starts_at value in bounds")
                .insert_any_distinct(attr.type_id(), range);
            self.attrs
                .ends_at
                .get_mut(end_idx)
                .expect("has initial ends_at value in bounds")
                .insert_any_distinct(attr.type_id(), range);
            self.attrs.ranges.insert_any_distinct(attr.type_id(), range);
            self.attrs
                .values
                .entry(range)
                .or_default()
                .insert_any_attribute(attr);
        }
    }

    /// Get a reference to the l l line's ll tokens.
    pub fn ll_tokens(&self) -> &[LLToken] {
        &self.ll_tokens
    }
}

impl LLLineAttrs {
    fn insert<T: 'static + std::fmt::Debug>(&mut self, range: LRange, value: T) {
        self.starts_at
            .get_mut(range.0)
            .expect("has initial starts_at value in bounds")
            .insert_distinct::<T>(range);
        self.ends_at
            .get_mut(range.1)
            .expect("has initial ends_at value in bounds")
            .insert_distinct::<T>(range);
        self.ranges.insert_distinct::<T>(range);
        self.values.entry(range).or_default().insert(value);
    }
}

#[track_caller]
fn assert_ll_lines_equals(first: &Rc<LLLine>, second: &Rc<LLLine>) {
    if !Rc::ptr_eq(first, second) {
        panic!("Two different lines used")
    }
}

// TODO rename
#[derive(Debug)]
pub struct LLCursorAssignment<Attr> {
    // private
    start_idx: usize,
    end_idx: usize,
    // provided from resolver
    value: Attr,
}

pub trait Resolver {
    type Attr: std::fmt::Debug + 'static;
    fn go(&self, selection: LLSelection) -> Vec<LLCursorAssignment<Self::Attr>>;
}
