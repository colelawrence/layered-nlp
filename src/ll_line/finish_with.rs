use super::{LLCursorAssignment, LLSelection};

pub trait FinishWith<T> {
    fn finish_with<Attr, F: Fn(T) -> Attr>(self, f: F) -> Vec<LLCursorAssignment<Attr>>;
}

impl<T, I: IntoIterator<Item = (LLSelection, T)>> FinishWith<T> for I {
    fn finish_with<Attr, F: Fn(T) -> Attr>(self, f: F) -> Vec<LLCursorAssignment<Attr>> {
        self.into_iter()
            .map(|(selection, t)| selection.finish_with_attr(f(t)))
            .collect()
    }
}
