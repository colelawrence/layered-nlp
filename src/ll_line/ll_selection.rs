use super::x::{XBackwards, XForwards};
use super::{assert_ll_lines_equals, LLCursorAssignment, LLLine, Rc, XMatch};

// Cutting Selection (selection) -> Iter<selection>
//  - split_by_x             :: [aaaxaaaxaaa] -> [aaa]x[aaa]x[aaa]
//  - find_by_x              :: [aaaxaaaxaaa] -> aaa[x]aaa[x]aaa
// skips overlapping
//  - split_by_forwards_x    :: [aaaxaaaxaaa] -> [aaa]x[aaa]x[aaa]
//  - find_by_forwards_x     :: [aaaxaaaxaaa] -> aaa[x]aaa[x]aaa
// skips overlapping
//  - split_by_backwards_x   :: [aaaxaaaxaaa] -> [aaa]x[aaa]x[aaa]
//  - find_by_backwards_x    :: [aaaxaaaxaaa] -> aaa[x]aaa[x]aaa
// Expand Selection (selection) -> Iter<selection>
//  - match_forwards_x          :: x[aaa]x -> x[aaax]
//  - match_backwards_x         :: x[aaa]x -> [xaaa]x
// Shrinking Selection (selection) -> Option<selection>
//  - trim_x                    :: [xaaax] -> x[aaa]x
//  - trim_leading_x            :: [xaaax] -> x[aaax]
//  - trim_trailing_x           :: [xaaax] -> [xaaa]x

// x
// start large, then test

/// Selections will never be empty
#[derive(Clone)]
pub struct LLSelection {
    pub(super) ll_line: Rc<LLLine>,
    /// Where to begin in the line (inclusive, default is 0)
    pub(super) start_idx: usize,
    /// Where to end in the line (inclusive, default is last idx)
    pub(super) end_idx: usize,
}

impl std::fmt::Debug for LLSelection {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("LLSelection")
            .field("start_idx", &self.start_idx)
            .field("end_idx", &self.end_idx)
            .finish()
    }
}

impl PartialEq for LLSelection {
    fn eq(&self, other: &Self) -> bool {
        self.start_idx == other.start_idx
            && self.end_idx == other.end_idx
            && Rc::ptr_eq(&self.ll_line, &other.ll_line)
    }
}

impl LLSelection {
    /// Returns None if the line is empty
    pub fn from_line(ll_line: Rc<LLLine>) -> Option<Self> {
        let line_len = ll_line.ll_tokens.len();
        if line_len > 0 {
            Some(LLSelection {
                ll_line,
                end_idx: line_len - 1,
                start_idx: 0,
            })
        } else {
            None // empty selection
        }
    }

    pub fn split_by<'a, M: XMatch<'a>>(&'a self, matcher: &M) -> Vec<LLSelection> {
        let matches = self.find_by(matcher);

        if matches.is_empty() {
            return vec![self.clone()];
        }

        if matches.len() == 1 {
            return matches[0]
                .0
                .start_idx
                .checked_sub(1)
                .and_then(|end_idx| self.selection_from(self.start_idx, end_idx))
                .into_iter()
                .chain(self.selection_from(matches[0].0.end_idx + 1, self.end_idx))
                .collect();
        }

        let start_opt = matches[0]
            .0
            .start_idx
            .checked_sub(1)
            .and_then(|end_idx| self.selection_from(self.start_idx, end_idx));

        let end_opt = self.selection_from(matches.last().unwrap().0.end_idx + 1, self.end_idx);
        start_opt
            .into_iter()
            .chain(matches.windows(2).into_iter().filter_map(|m| {
                m[1].0
                    .start_idx
                    .checked_sub(1)
                    .and_then(|end_idx| self.selection_from(m[0].0.end_idx + 1, end_idx))
            }))
            .chain(end_opt)
            .collect()
    }

    pub fn find_by<'a, M: XMatch<'a>>(&'a self, matcher: &M) -> Vec<(LLSelection, M::Out)> {
        (self.start_idx..=self.end_idx)
            .map(|i| {
                let forwards = XForwards { from_idx: i };

                matcher
                    .go(&forwards, &self.ll_line)
                    .into_iter()
                    .map(move |(out, next_idx)| {
                        (
                            LLSelection {
                                start_idx: i,
                                end_idx: next_idx.0,
                                ll_line: self.ll_line.clone(),
                            },
                            out,
                        )
                    })
            })
            .flatten()
            .collect()
    }

    pub fn find_first_by<'a, M: XMatch<'a>>(
        &'a self,
        matcher: &M,
    ) -> Option<(LLSelection, M::Out)> {
        self.find_by(matcher).into_iter().next()
    }

    pub fn find_by_forwards_and_backwards<'a, M: XMatch<'a>>(
        &'a self,
        matcher: &M,
    ) -> Vec<(LLSelection, M::Out)> {
        (self.start_idx..=self.end_idx)
            .map(|i| {
                let forwards = XForwards { from_idx: i };

                matcher
                    .go(&forwards, &self.ll_line)
                    .into_iter()
                    .map(move |(out, next_idx)| {
                        (
                            LLSelection {
                                start_idx: i,
                                end_idx: next_idx.0,
                                ll_line: self.ll_line.clone(),
                            },
                            out,
                        )
                    })
                    .chain({
                        let backwards = XBackwards { from_idx: i };

                        matcher.go(&backwards, &self.ll_line).into_iter().map(
                            move |(out, next_idx)| {
                                (
                                    LLSelection {
                                        start_idx: next_idx.0,
                                        end_idx: i,
                                        ll_line: self.ll_line.clone(),
                                    },
                                    out,
                                )
                            },
                        )
                    })
            })
            .flatten()
            .collect()
    }

    pub fn match_forwards<'a, M: XMatch<'a>>(&'a self, matcher: &M) -> Vec<(LLSelection, M::Out)> {
        // [ ... ] - Current selection
        //        [ ... ] - Trying to match Attr
        if self.end_idx + 1 == self.ll_line.ll_tokens.len() {
            return Vec::new();
        }

        let forwards = XForwards {
            from_idx: self.end_idx + 1,
        };

        matcher
            .go(&forwards, &self.ll_line)
            .into_iter()
            .map(|(out, next_idx)| {
                (
                    LLSelection {
                        start_idx: self.start_idx,
                        end_idx: next_idx.0,
                        ll_line: self.ll_line.clone(),
                    },
                    out,
                )
            })
            .collect()
    }

    pub fn match_first_forwards<'a, M: XMatch<'a>>(
        &'a self,
        matcher: &M,
    ) -> Option<(LLSelection, M::Out)> {
        self.match_forwards(matcher).into_iter().next()
    }

    pub fn match_forwards_longest<'a, M: XMatch<'a>>(
        &'a self,
        _matcher: &M,
    ) -> Option<(LLSelection, M::Out)> {
        todo!()
    }

    pub fn match_forwards_shortest<'a, M: XMatch<'a>>(
        &'a self,
        _matcher: &M,
    ) -> Option<(LLSelection, M::Out)> {
        todo!()
    }

    pub fn match_backwards<'a, M: XMatch<'a>>(&'a self, matcher: &M) -> Vec<(LLSelection, M::Out)> {
        if self.start_idx == 0 {
            return Vec::new();
        }

        let backwards = XBackwards {
            from_idx: self.start_idx - 1,
        };

        matcher
            .go(&backwards, &self.ll_line)
            .into_iter()
            .map(|(out, next_idx)| {
                (
                    LLSelection {
                        start_idx: next_idx.0,
                        end_idx: self.end_idx,
                        ll_line: self.ll_line.clone(),
                    },
                    out,
                )
            })
            .collect()
    }

    pub fn match_first_backwards<'a, M: XMatch<'a>>(
        &'a self,
        matcher: &M,
    ) -> Option<(LLSelection, M::Out)> {
        self.match_backwards(matcher).into_iter().next()
    }

    pub fn after(&self) -> Option<LLSelection> {
        let ll_line_end = self.ll_line.ll_tokens.len() - 1;

        if self.start_idx + 1 == ll_line_end {
            None
        } else {
            Some(LLSelection {
                ll_line: self.ll_line.clone(),
                start_idx: self.end_idx + 1,
                end_idx: ll_line_end,
            })
        }
    }

    fn selection_from(&self, mut start_idx: usize, mut end_idx: usize) -> Option<LLSelection> {
        start_idx = start_idx.max(self.start_idx);
        end_idx = end_idx.min(self.end_idx);
        if start_idx <= end_idx {
            Some(LLSelection {
                ll_line: self.ll_line.clone(),
                end_idx,
                start_idx,
            })
        } else {
            None
        }
    }

    // Hmmm... TODO: Unit test this thoroughly
    pub fn split_with(&self, other_selection: &LLSelection) -> [Option<LLSelection>; 2] {
        assert_ll_lines_equals(&self.ll_line, &other_selection.ll_line);

        // 1   [          ]
        // 2                   [      ]
        //     [          ]
        //               ]      [
        // 1   [          ]
        // 2           [      ]
        //     [      ]

        [
            if other_selection.start_idx > 0 {
                // underflow protection
                self.selection_from(self.start_idx, other_selection.start_idx - 1)
            } else {
                None
            },
            self.selection_from(other_selection.end_idx + 1, self.end_idx),
        ]
    }

    pub fn finish_with_attr<Attr>(&self, value: Attr) -> LLCursorAssignment<Attr> {
        LLCursorAssignment {
            end_idx: self.end_idx,
            start_idx: self.start_idx,
            value,
        }
    }
}
