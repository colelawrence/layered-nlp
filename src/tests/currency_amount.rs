use crate::ll_line::{x, LLSelection};

use super::*;

#[derive(Clone, Debug)]
enum CurrencySymbol {
    Euro,
    USDDollars,
    Yen,
}

struct CurrencySymbolResolver;

impl Resolver for CurrencySymbolResolver {
    type Attr = CurrencySymbol;

    fn go(&self, selection: LLSelection) -> Vec<LLCursorAssignment<Self::Attr>> {
        selection
            .find_by(&x::token_has_any(&['€', '$', '¥']))
            .into_iter()
            .map(|(sel, sym_ch)| match sym_ch {
                '$' => sel.finish_with_attr(CurrencySymbol::USDDollars),
                '€' => sel.finish_with_attr(CurrencySymbol::Euro),
                '¥' => sel.finish_with_attr(CurrencySymbol::Yen),
                _ => unreachable!(),
            })
            .collect()
    }
}

#[derive(Clone, Debug)]
struct Amount(Decimal);

struct AmountResolver {
    /// Configure for localization
    delimiters: Vec<char>,
    decimal: char,
}

impl Resolver for AmountResolver {
    type Attr = Amount;

    fn go(&self, mut selection: LLSelection) -> Vec<LLCursorAssignment<Self::Attr>> {
        let mut attrs = vec![];

        while let Some((mut sel, (_, text))) =
            selection.find_first_by(&(x::attr_eq(&TextTag::NATN), x::token_text()))
        {
            let mut number_string = String::from(text);

            // Could we have an enum here perhaps something like:
            // enum {
            //   NaturalNumber(?),
            //   TrailingDelimiter(?),
            //   TrailingDecimal(?)
            // }
            let mut last_valid_cursor = None;

            // Avoid trailing delimeters
            loop {
                if let Some((delimeter_sel, _)) =
                    // extend_forwards?
                // extend_if_hugging_forwards?
                    sel.match_first_forwards(&x::token_has_any(self.delimiters.as_slice()))
                {
                    last_valid_cursor = Some(sel);
                    sel = delimeter_sel;
                } else {
                    break;
                }

                if let Some((following_delimeter_cur, (_, text))) =
                    sel.match_first_forwards(&(x::attr_eq(&TextTag::NATN), x::token_text()))
                {
                    number_string.push_str(text);
                    last_valid_cursor = None;
                    sel = following_delimeter_cur;
                } else {
                    break;
                }
            }

            if last_valid_cursor.is_none() {
                if let Some((with_decimal_sel, _)) =
                    sel.match_first_forwards(&x::token_has_any(&[self.decimal]))
                {
                    last_valid_cursor = Some(sel);
                    number_string.push('.');
                    sel = with_decimal_sel;
                }

                if let Some((following_decimal_cur, ((), text))) =
                    sel.match_first_forwards(&(x::attr_eq(&TextTag::NATN), x::token_text()))
                {
                    number_string.push_str(text);
                    last_valid_cursor = None;
                    sel = following_decimal_cur;
                }
            }

            // dbg!(number_string.clone());

            attrs.push(
                last_valid_cursor
                    .unwrap_or_else(|| sel.clone())
                    .finish_with_attr(Amount(dbg!(number_string).parse::<Decimal>().unwrap())),
            );

            if let [_, Some(right_sel)] = selection.split_with(&sel) {
                selection = right_sel;
            } else {
                break;
            }
        }

        attrs
    }
}

#[derive(Debug, Clone)]
struct CurrencyAmount(CurrencySymbol, Amount);

struct CurrencyAmountResolver;

impl Resolver for CurrencyAmountResolver {
    type Attr = CurrencyAmount;

    fn go(&self, selection: LLSelection) -> Vec<LLCursorAssignment<Self::Attr>> {
        selection
            .find_by(&x::attr::<CurrencySymbol>())
            .into_iter()
            .filter_map(|(sel, curr_sym)| {
                sel.match_first_forwards(&x::attr::<Amount>())
                    .or_else(|| sel.match_first_backwards(&x::attr::<Amount>()))
                    .map(|(cur_with_amt, amt)| {
                        cur_with_amt.finish_with_attr(CurrencyAmount(curr_sym.clone(), amt.clone()))
                    })
            })
            .collect()
    }
}

#[test]
fn amount_resolver_foo() {
    let ll_line = test_line("100,.000").run(&AmountResolver {
        delimiters: vec![','],
        decimal: '.',
    });

    let mut ll_line_display = LLLineDisplay::new(&ll_line);
    ll_line_display.include::<Amount>();

    insta::assert_display_snapshot!(ll_line_display, @r###"
    100  ,  .  000
    ╰─╯Amount(100)
               ╰─╯Amount(0)
    "###);
}

#[test]
fn amount_resolver_bar() {
    let ll_line = test_line("10.00,00").run(&AmountResolver {
        delimiters: vec![','],
        decimal: '.',
    });

    let mut ll_line_display = LLLineDisplay::new(&ll_line);
    ll_line_display.include::<Amount>();

    insta::assert_display_snapshot!(ll_line_display, @r###"
    10  .  00  ,  00
    ╰───────╯Amount(10.00)
                  ╰╯Amount(0)
    "###);
}

#[test]
fn it_works_usd() {
    let ll_line = test_line("$1,000.25")
        .run(&CurrencySymbolResolver)
        .run(&AmountResolver {
            delimiters: vec![','],
            decimal: '.',
        })
        .run(&CurrencyAmountResolver);

    let mut ll_line_display = LLLineDisplay::new(&ll_line);
    ll_line_display.include::<CurrencySymbol>();
    ll_line_display.include::<Amount>();
    ll_line_display.include::<CurrencyAmount>();

    insta::assert_display_snapshot!(ll_line_display, @r###"
    $  1  ,  000  .  25
    ╰USDDollars
       ╰──────────────╯Amount(1000.25)
    ╰─────────────────╯CurrencyAmount(USDDollars, Amount(1000.25))
    "###);
}
#[test]
fn it_works_euro() {
    let ll_line = test_line(". 1 000,25€")
        .run(&CurrencySymbolResolver)
        .run(&AmountResolver {
            delimiters: vec![' '],
            decimal: ',',
        })
        .run(&CurrencyAmountResolver);

    let mut ll_line_display = LLLineDisplay::new(&ll_line);
    ll_line_display.include::<CurrencySymbol>();
    ll_line_display.include::<Amount>();
    ll_line_display.include::<CurrencyAmount>();

    insta::assert_display_snapshot!(ll_line_display, @r###"
    .     1     000  ,  25  €
                            ╰Euro
          ╰──────────────╯Amount(1000.25)
          ╰─────────────────╯CurrencyAmount(Euro, Amount(1000.25))
    "###);
}

#[test]
fn it_works_yen() {
    let ll_line = test_line("¥227")
        .run(&CurrencySymbolResolver)
        .run(&AmountResolver {
            delimiters: vec![','],
            // TODO: What if a currency does not have decimals?
            decimal: ' ',
        })
        .run(&CurrencyAmountResolver);

    let mut ll_line_display = LLLineDisplay::new(&ll_line);
    ll_line_display.include::<CurrencySymbol>();
    ll_line_display.include::<Amount>();
    ll_line_display.include::<CurrencyAmount>();

    insta::assert_display_snapshot!(ll_line_display, @r###"
    ¥  227
    ╰Yen
       ╰─╯Amount(227)
    ╰────╯CurrencyAmount(Yen, Amount(227))
    "###);
}

#[derive(Clone, Debug)]
enum WillCaresSymbol {
    Tests,
}

struct WillCaresResolver;

impl Resolver for WillCaresResolver {
    type Attr = WillCaresSymbol;

    fn go(&self, selection: LLSelection) -> Vec<LLCursorAssignment<Self::Attr>> {
        selection
            .find_by(&(x::attr_eq(&TextTag::WORD), x::token_text()))
            // .find_by(&x::token_has_any(&["hens"]))
            .into_iter()
            .map(|(sel, (_, sym_ch))| match sym_ch {
                "tests" => sel.finish_with_attr(WillCaresSymbol::Tests),
                _ => unreachable!(),
            })
            .collect()
    }
}

#[derive(Debug, Clone)]
struct WillCaresAmount(WillCaresSymbol, Amount);

struct WillCaresAmountResolver;

impl Resolver for WillCaresAmountResolver {
    type Attr = WillCaresAmount;

    fn go(&self, selection: LLSelection) -> Vec<LLCursorAssignment<Self::Attr>> {
        // find_by(Symbol AND Amount OR Amount AND Symbol)
        selection
            .find_by(&x::attr::<WillCaresSymbol>())
            .into_iter()
            .filter_map(|(sel, will_cares_sym)| {
                // How about sel.match_first(Prefer::Forwards)?
                sel.match_first(crate::ll_line::Direction::Forwards, &x::attr::<Amount>())
                    .map(|(will_cares_with_amt, amt)| {
                        will_cares_with_amt
                            .finish_with_attr(WillCaresAmount(will_cares_sym.clone(), amt.clone()))
                    })
            })
            .collect()
    }
}

#[test]
fn it_works_for_things_that_will_cares_about() {
    let ll_line = test_line("1337tests")
        .run(&WillCaresResolver)
        .run(&AmountResolver {
            delimiters: vec![','],
            // TODO: What if a currency does not have decimals?
            decimal: ' ',
        })
        .run(&WillCaresAmountResolver);

    let mut ll_line_display = LLLineDisplay::new(&ll_line);
    ll_line_display.include::<WillCaresSymbol>();
    ll_line_display.include::<Amount>();
    ll_line_display.include::<WillCaresAmount>();

    insta::assert_display_snapshot!(ll_line_display, @r###"
    1337  tests
          ╰───╯Tests
    ╰──╯Amount(1337)
    ╰─────────╯WillCaresAmount(Tests, Amount(1337))
    "###);
}
