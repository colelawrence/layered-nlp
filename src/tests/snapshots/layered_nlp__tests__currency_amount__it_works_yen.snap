---
source: src/tests/currency_amount.rs
expression: ll_line_display

---
¥  227
╰Yen
   ╰─╯Amount(227)
╰────╯CurrencyAmount(Yen, Amount(227))
